File: kotlin-orders/kotlin-orders-dto/src/main/java/com/mvv/bank/orders/rest/conversion/CashOrderMapper.kt
package com.mvv.bank.orders.rest.conversion

import com.mvv.bank.orders.conversion.DomainPrimitiveMappers
import com.mvv.bank.orders.conversion.MAP_STRUCT_COMPONENT_MODEL
import com.mvv.bank.orders.conversion.GenericOrderDtoDomainConversion

import com.mvv.bank.orders.domain.CashMarketOrder as DomainMarketOrder
import com.mvv.bank.orders.domain.CashLimitOrder as DomainLimitOrder
import com.mvv.bank.orders.domain.AbstractCashOrder as DomainOrder
import com.mvv.bank.orders.domain.CashStopOrder as DomainStopOrder
import com.mvv.bank.orders.domain.OrderType as DomainOrderType

import com.mvv.bank.orders.rest.entities.OrderType as DtoOrderType
import com.mvv.bank.orders.rest.entities.BaseOrder as DtoBaseOrder
import com.mvv.bank.orders.rest.entities.CashOrder as DtoOrder

import org.mapstruct.*
import kotlin.reflect.KClass


@Mapper(
    componentModel = MAP_STRUCT_COMPONENT_MODEL,
    config = DomainPrimitiveMappers::class,
    uses = [
        FxRateMapper::class,
        AmountMapper::class,
    ]
)
@Suppress("CdiInjectionPointsInspection")
abstract class CashOrderMapper: AbstractRestOrderMapper(),
    GenericOrderDtoDomainConversion<DtoBaseOrder, DtoOrder, DtoOrderType, DomainOrder, DomainMarketOrder, DomainLimitOrder, DomainStopOrder> {

    // Temporary overriding because MapStruct generates method twice in case of diamond inheritance.
    // T O D O: remove it when it is fixed in MapStruct
    abstract override fun orderTypeToDomain(source: DtoOrderType): DomainOrderType


    override fun chooseOrderTypeClass(orderType: DomainOrderType): KClass<*> = orderType.cashDomainType

    // ***************************************************************************************************

    // to avoid warnings
    @Mapping(target = "limitPrice", ignore = true)
    @Mapping(target = "stopPrice", ignore = true)
    @Mapping(target = "dailyExecutionType", ignore = true)
    abstract fun baseOrderAttrsToDto(source: DomainOrder, @MappingTarget target: DtoOrder): DtoOrder

    @InheritConfiguration(name = "baseOrderAttrsToDto")
    // because earlier it was marked as ignored
    @Mapping(source = "limitPrice", target = "limitPrice")
    @Mapping(source = "dailyExecutionType", target = "dailyExecutionType")
    abstract override fun limitOrderToDto(source: DomainLimitOrder): DtoOrder

    @InheritConfiguration(name = "baseOrderAttrsToDto")
    // because earlier it was marked as ignored
    @Mapping(source = "stopPrice", target = "stopPrice")
    @Mapping(source = "dailyExecutionType", target = "dailyExecutionType")
    abstract override fun stopOrderToDto(source: DomainStopOrder): DtoOrder

    @InheritConfiguration(name = "baseOrderAttrsToDto")
    abstract override fun marketOrderToDto(source: DomainMarketOrder): DtoOrder


    // ***************************************************************************************************

    @Mapping(target = "product", ignore = true)
    @Mapping(target = "resultingPrice", ignore = true)
    @Mapping(target = "resultingQuote", ignore = true)
    abstract fun baseOrderAttrsToDomain(source: DtoOrder, @MappingTarget target: DomainOrder): DomainOrder

    @InheritConfiguration(name = "baseOrderAttrsToDomain")
    abstract override fun dtoToLimitOrder(source: DtoOrder): DomainLimitOrder

    @InheritConfiguration(name = "baseOrderAttrsToDomain")
    abstract override fun dtoToStopOrder(source: DtoOrder): DomainStopOrder

    @InheritConfiguration(name = "baseOrderAttrsToDomain")
    abstract override fun dtoToMarketOrder(source: DtoOrder): DomainMarketOrder
}

File: scala-orders/src/main/scala/com/mvv/bank/orders/rest/conversion/CashOrderMapper.scala
package com.mvv.bank.orders.rest.conversion

import org.mapstruct.{
  Mapper, Mapping, ObjectFactory, BeanMapping, MappingTarget, InheritConfiguration, BeforeMapping
}
import com.mvv.bank.orders.domain.{
  createCashOrderFor as createDomainOrderFor,
  OrderType         as DomainOrderType,
  AbstractCashOrder as DomainBaseOrder,
  CashLimitOrder    as DomainLimitOrder,
  CashMarketOrder   as DomainMarketOrder,
  CashStopOrder     as DomainStopOrder,
}
import com.mvv.bank.orders.rest.entities.{
  BaseOrder as DtoBaseOrder,
  CashOrder as DtoCashOrder,
  OrderType as DtoOrderType,
}
import com.mvv.log.{ safe, underlyingSafe }
import com.mvv.nullables.{ isNotNull, ifNull }
import com.mvv.utils.require
import com.mvv.scala.mapstruct.getScalaMapStructMapper

import scala.annotation.unused



abstract class CashOrderMapper extends AbstractOrderMapper, Cloneable :

  /*  @InheritConfiguration is not used because now it allows ONLY to specify FULL configuration.
   *  If we write
   *   {{{
   *    @Mapping(target = "stopPrice", ignore = true)
   *    @Mapping(target = "limitPrice", ignore = true)
   *    @Mapping(target = "dailyExecutionType", ignore = true)
   *    @BeanMapping(ignoreUnmappedSourceProperties = Array("log", "product", "resultingQuote"))
   *    def baseOrderAttrsToDto(source: DomainBaseOrder, @MappingTarget target: DtoCashOrder): Unit
   *   }}}
   *  We will need to repeat needed fields manually and @BeanMapping is not inherited
   *  (it just kills main use of MapStruct :-( )
   *   {{{
   *    @Mapping(target = "limitPrice", source = "limitPrice")                  // ?? manually again ??
   *    @Mapping(target = "dailyExecutionType", source = "dailyExecutionType")  // ?? manually again ??
   *    @InheritConfiguration("baseOrderAttrsToDto")
   *    @BeanMapping(ignoreUnmappedSourceProperties = Array("log", "product", "resultingQuote")) // Oops ?? It is not inherited !! Piece of sh!!!
   *    def limitOrderToDto(source: DomainLimitOrder): DtoCashOrder
   *   }}}
   *
   *
   *  If we use method with default empty body as template it is just not ignored at all :-(
   *   {{{
   *    @BeanMapping(ignoreUnmappedSourceProperties = Array("log", "product", "resultingQuote"))
   *    def baseOrderAttrsToDto(source: DomainLimitOrder, @MappingTarget target: DtoCashOrder): Unit = {}
   *   }}}
   */

  @Mapping(target = "stopPrice", ignore = true)
  @Mapping(target = "limitPrice", ignore = true)
  @Mapping(target = "dailyExecutionType", ignore = true)
  @BeanMapping(ignoreUnmappedSourceProperties = Array("log", "product", "resultingQuote"))
  def baseOrderAttrsToDto(source: DomainBaseOrder, @MappingTarget target: DtoCashOrder): Unit

  @Mapping(target = "limitPrice", source = "limitPrice")                  // ?? manually again ??
  @Mapping(target = "dailyExecutionType", source = "dailyExecutionType")  // ?? manually again ??
  @InheritConfiguration("baseOrderAttrsToDto")
  // Oops ?? It is not inherited !! Piece of sh!!!
  // see bug https://github.com/mapstruct/mapstruct/issues/2092
  @BeanMapping(ignoreUnmappedSourceProperties = Array("log", "product", "resultingQuote"))
  def limitOrderToDto(source: DomainLimitOrder): DtoCashOrder

  @Mapping(target = "stopPrice", source = "stopPrice")                   // ?? manually again ??
  @Mapping(target = "dailyExecutionType", source = "dailyExecutionType") // ?? manually again ??
  @InheritConfiguration("baseOrderAttrsToDto")
  // Oops ?? It is not inherited !! Piece of sh!!!
  // see bug https://github.com/mapstruct/mapstruct/issues/2092
  @BeanMapping(ignoreUnmappedSourceProperties = Array("log", "product", "resultingQuote"))
  def stopOrderToDto(source: DomainStopOrder): DtoCashOrder

  @InheritConfiguration("baseOrderAttrsToDto")
  // Oops ?? It is not inherited !! Piece of sh!!!
  // see bug https://github.com/mapstruct/mapstruct/issues/2092
  @BeanMapping(ignoreUnmappedSourceProperties = Array("log", "product", "resultingQuote"))
  def marketOrderToDto(source: DomainMarketOrder): DtoCashOrder

  def toDto(source: DomainBaseOrder): DtoCashOrder =
    source match
      case stopOrder: DomainStopOrder => stopOrderToDto(stopOrder)
      case limitOrder: DomainLimitOrder => limitOrderToDto(limitOrder)
      case marketOrder: DomainMarketOrder => marketOrderToDto(marketOrder)



  // ***************************************************************************************************

  @Mapping(target = "log", ignore = true) // not property at all
  @Mapping(target = "product", ignore = true)
  @Mapping(target = "orderType", ignore = true)     // read-only
  @Mapping(target = "priceCurrency", ignore = true) // read-only
  @BeanMapping(ignoreUnmappedSourceProperties = Array(
    // not used for LimitOrder
    "stopPrice",
    // read-only
    "orderType", "priceCurrency",
  ))
  def limitOrderToDomain(source: DtoCashOrder): DomainLimitOrder

  @Mapping(target = "log", ignore = true) // not property at all
  @Mapping(target = "product", ignore = true)
  @Mapping(target = "orderType", ignore = true)     // read-only
  @Mapping(target = "priceCurrency", ignore = true) // read-only
  @BeanMapping(ignoreUnmappedSourceProperties = Array(
    // not used for StopOrder
    "limitPrice",
    // read-only
    "orderType", "priceCurrency"))
  def stopOrderToDomain(source: DtoCashOrder): DomainStopOrder

  @Mapping(target = "log", ignore = true) // not property at all
  @Mapping(target = "product", ignore = true)
  @Mapping(target = "orderType", ignore = true)     // read-only
  @Mapping(target = "priceCurrency", ignore = true) // read-only
  @BeanMapping(ignoreUnmappedSourceProperties = Array(
    // not used for MarketOrder
    "limitPrice", "stopPrice", "dailyExecutionType",
    // read-only
    "orderType", "priceCurrency",
  ))
  def marketOrderToDomain(source: DtoCashOrder): DomainMarketOrder

  def toDomain(source: DtoCashOrder): DomainBaseOrder =
    source.orderType match
      case DtoOrderType.STOP_ORDER   => stopOrderToDomain(source)
      case DtoOrderType.LIMIT_ORDER  => limitOrderToDomain(source)
      case DtoOrderType.MARKET_ORDER => marketOrderToDomain(source)


  @BeforeMapping
  //noinspection ScalaUnusedSymbol
  def validateDomainOrderBeforeConverting(source: DomainBaseOrder, @MappingTarget target: DtoBaseOrder): Unit =
    // We do not perform strict validation to have possibility to see/analyze probably wrong order.
    // source.validateCurrentState()

    // But we will use simple validation to have better error message.
    // (error/exception will be thrown in any case)
    source.checkRequiredPropsAreInitialized()


  @BeforeMapping
  //noinspection ScalaUnusedSymbol
  def validateDtoOrderBeforeConverting(source: DtoBaseOrder, @MappingTarget target: Any): Unit =
    if source.orderType == DtoOrderType.MARKET_ORDER then
      require(source.limitPrice.isEmpty && source.stopPrice.isEmpty,
        s"Market price cannot have limit/stop price (${source.limitPrice.underlyingSafe}/${source.stopPrice.underlyingSafe}).")
      require(source.dailyExecutionType.isEmpty,
        s"Market price cannot have daily execution type (${source.dailyExecutionType.underlyingSafe}).")


  @ObjectFactory  @unused
  def createDomainCashOrder(source: DtoBaseOrder): DomainBaseOrder =
    // T O D O: please make it easier (without injecting EnumMappers) if you know ho to do it...
    val _enumMappers: EnumMappers = enumMappers.ifNull( getScalaMapStructMapper[EnumMappers]() )
    createDomainOrderFor(_enumMappers.OrderTypeToDomain(source.orderType))

  override def clone(): CashOrderMapper = super.clone().asInstanceOf[CashOrderMapper]
